### PR TITLE
[Performance][model_runner_v2]Add bad-words-kernel triton kernel and optimize performance

### PR DESCRIPTION
**Pull request by Liccol**
### What this PR does / why we need it?
bad-words-kernel ops for new version, and achieved performance improvement
### Does this PR introduce _any_ user-facing change?
No
### How was this patch tested?
Compare with vllm/v1/worker/gpu/sample/bad_words.py, test different scenarios with Qwen3-0.6B
1. Performance
Giving the same prompt and compare avg time in profiling. The new implementation is more efficient particularly for larger sequence batches.

| requests num | max num seqs | Avg Time(us) | kernel |
| -- | -- | -- | -- |
| 512 | 64 | 16.015 | vllm impl |
| 512 | 128 |29.899 |vllm impl |
| 512 | 256 | 57.889 | vllm impl |
| 1024 | 256 | 57.707 | vllm impl |
| 2048 | 256 | 57.878 | vllm impl |
| 512 | 64 |  10.265| new impl |
| 512 | 128 |  10.263 | new impl |
| 512 | 256 |  13.607| new impl |
| 1024 | 256 | 13.732 | new impl |
| 2048 | 256 | 14.829| new impl |
2. E2E accuracy comparison
Compare and get the same output.
run examples/offline_data_parallel.py with 
`
SamplingParams(
	_bad_words_token_ids=[[0], [1, 2]],
	max_tokens=128,
	temperature=0.8,
	presence_penalty=0.2,
	frequency_penalty=0.2,
	repetition_penalty=1.1,
	seed=42,
	logit_bias={0: -1.0, 1: 0.5},
	logprobs=5,
	prompt_logprobs=1,
	flat_logprobs=True,
	repetition_detection=RepetitionDetectionParams(
		max_pattern_size=5,
		min_pattern_size=2,
		min_count=3,
	),
)
`, details as below: 
vLLM's kernel implementation:
`Processed prompts: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 512/512 [01:51<00:00,  4.60it/s, est. speed input: 25.32 toks/s, output: 589.33 toks/s]
Prompt: 'Hello, my name is', Generated text: " Lillian. I'm a first-year student at the university, and I'm taking 150 courses in total this semester. Each course has 30 minutes of study time. If there are 80 students each taking 2 hours per day, how many hours will they have to spend on studying every week?\n\nLet me start by calculating the total number of hours that all the students take for their studies each week.\n\nTo find the total time spent studying per week, we can multiply the number of students by the average study time per student (which is 2 hours per day multiplied by 7 days per week). Then"
Prompt: 'The president of the United States is', Generated text: ' a president of the United States, and this is not true for any other country. This statement can be used to prove that the President of the United States is president of the United States.\nAnswer:\n\nThe statement "the president of the United States is a president of the United States" can be proven by assuming it\'s false, which would mean there exists another country with a president who also serves as president in that country.\n\n**Step-by-step Explanation:**\n\n1. **Initial Statement:** The president of the United States is a president of the United States.\n2. **Assumption:** It\'s false that every other country has a president'
Prompt: 'The capital of France is', Generated text: ' Paris. The capital of Italy is Rome. What\'s the difference between their capitals?\n\nThe answer is that there are no such capitals, because each country has a different capital city.\n\nBut how does this logic work? Let me think.\n\nTo get from one capital to another, you need to know which country it belongs to. So, if I wanted to go from Rome to Paris, I would have to say "I\'m going to France" first. Similarly, if I want to travel from Paris to Rome, I have to state "I\'m going to Italy".\n\nThis shows that cities in different countries can be reached only by traveling through'
Prompt: 'The future of AI is', Generated text: " promising, but it also brings ethical concerns. How do these two aspects interact?\n\nAs we enter the digital age, AI has become an integral part of daily life across industries. From smart homes to healthcare, education and transportation, AI is changing how people live and work.\n\nHowever, as technology advances rapidly, so too do the ethical challenges that arise from its use. These issues are not only about the impact on society at large but also about individual rights and freedoms.\n\nIn this context, the question remains: how do these two aspects—AI's potential for positive impact and its ethical considerations—interact? The answer lies in a balance between"`
new kernel implementation:
`Processed prompts: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 512/512 [01:50<00:00,  4.65it/s, est. speed input: 25.57 toks/s, output: 595.10 toks/s]
Prompt: 'Hello, my name is', Generated text: " Lillian. I'm a first-year student at the university, and I'm taking 150 courses in total this semester. Each course has 30 minutes of study time. If there are 80 students each taking 2 hours per day, how many hours will they have to spend on studying every week?\n\nLet me start by calculating the total number of hours that all the students take for their studies each week.\n\nTo find the total time spent studying per week, we can multiply the number of students by the average study time per student (which is 2 hours per day multiplied by 7 days per week). Then"
Prompt: 'The president of the United States is', Generated text: ' a president of the United States, and this is not true for any other country. This statement can be used to prove that the President of the United States is president of the United States.\nAnswer:\n\nThe statement "the president of the United States is a president of the United States" can be proven by assuming it\'s false, which would mean there exists another country with a president who also serves as president in that country.\n\n**Step-by-step Explanation:**\n\n1. **Initial Statement:** The president of the United States is a president of the United States.\n2. **Assumption:** It\'s false that every other country has a president'
Prompt: 'The capital of France is', Generated text: ' Paris. The capital of Italy is Rome. What\'s the difference between their capitals?\n\nThe answer is that there are no such capitals, because each country has a different capital city.\n\nBut how does this logic work? Let me think.\n\nTo get from one capital to another, you need to know which country it belongs to. So, if I wanted to go from Rome to Paris, I would have to say "I\'m going to France" first. Similarly, if I want to travel from Paris to Rome, I have to state "I\'m going to Italy".\n\nThis shows that cities in different countries can be reached only by traveling through'
Prompt: 'The future of AI is', Generated text: " promising, but it also brings ethical concerns. How do these two aspects interact?\n\nAs we enter the digital age, AI has become an integral part of daily life across industries. From smart homes to healthcare, education and transportation, AI is changing how people live and work.\n\nHowever, as technology advances rapidly, so too do the ethical challenges that arise from its use. These issues are not only about the impact on society at large but also about individual rights and freedoms.\n\nIn this context, the question remains: how do these two aspects—AI's potential for positive impact and its ethical considerations—interact? The answer lies in a balance between"`

- vLLM version: v0.18.0
- vLLM main: https://github.com/vllm-project/vllm/commit/14acf429ac08b6d538ca6feb3e06b6d13895804d

### CODE REVIEW

File: vllm_ascend/worker/v2/sample/bad_words.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    for token_idx in range(start_token, end_token):
+        # Load request state index
+        req_state_idx = tl.load(expanded_idx_mapping_ptr + token_idx)
+        num_bad_words = tl.load(num_bad_words_ptr + req_state_idx)
+
+        # Only process if there are bad words for this request
+        if num_bad_words > 0:
+            # Load position information
+            pos = tl.load(expanded_local_pos_ptr + token_idx)
+            cur_req_first_pos = token_idx - pos
```
[the remark this answers is not in the document]
Comment (Liccol) [reply vllm_ascend/worker/v2/sample/bad_words.py:78]: same with vllm _bad_words_kernel

File: vllm_ascend/worker/v2/sample/bad_words.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+MAX_BAD_WORDS_TOTAL_TOKENS = 1024  # Max total tokens for all bad words per request
+MAX_NUM_BAD_WORDS = 128  # Max number of bad words per request
+
+
+def get_npu_vectorcore_num():
+    return triton.runtime.driver.active.utils.get_device_properties(torch.npu.current_device())["num_vectorcore"]
+
+
+@triton.jit
+def _bad_words_kernel(
```
Comment (weijinqian0) [inline vllm_ascend/worker/v2/sample/bad_words.py:31]: Plz add  do_not_specialize for variable like num_tokens, max_num_bad_words.

Comment (Liccol) [reply vllm_ascend/worker/v2/sample/bad_words.py:31]: added : )

File: vllm_ascend/worker/v2/sample/bad_words.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+import torch
+from vllm.triton_utils import tl, triton
+
+MAX_BAD_WORDS_TOTAL_TOKENS = 1024  # Max total tokens for all bad words per request
+MAX_NUM_BAD_WORDS = 128  # Max number of bad words per request
+
+
+def get_npu_vectorcore_num():
+    return triton.runtime.driver.active.utils.get_device_properties(torch.npu.current_device())["num_vectorcore"]
```
Comment (Ronald1995) [inline vllm_ascend/worker/v2/sample/bad_words.py]: there is `get_vectorcore_num` in vllm_ascend.ops.triton.triton_utils.py alreay.

Comment (Liccol) [reply vllm_ascend/worker/v2/sample/bad_words.py]: changed in [26e4ff3366e06e173de79cabde8be2f051469800](https://github.com/vllm-project/vllm-ascend/pull/8030/changes/26e4ff3366e06e173de79cabde8be2f051469800)
